Fix completion of Java packages/classes on Windows.

peterwang opened the pull request:
Since file pathes inside jar files contain '/' instead
of `File/separator', '/' also should be replaced with '.'
Before this fix, completion of java packages such as
java.util.*, java.nio.* does not work, it works now.

vemv reviewed (commented):
(contributing a review as someone who helps maintaining cider-nrepl)

I'd add a comment and deftests so that maintainers won't accidentally undo this bugfix in a future.

It also would help to expand the PR description with examples of the problematic strings.

We also should be sure that macOS/Linux users won't be affected (we'd gain that confidence with detailed testing and PR desc)

Cheers - V

peterwang commented:
Just checked the existing tests, string prefix `java.util` had been already covered. 
I guess that no tests had been running on Windows previously, so the issue was not caught.
BTW, I just ran `lein test` on windows: before the fix, there are 13 FAILED tests, after that, there are 6 FAILED tests. It is actually getting better. Currently, I have no time to further fix the 6 FAILED tests on Windows. Since all tests still run on Linux after the fix, I think is should be safe to merge the fix.

vemv commented:
Hi, would you consider addressing the simplest things I suggested? Detailed PR desc + inline comment.

Also if you post the output of the failing tests we can try solving them. I don't have a windows machine and setting that up in CI is quite hard (which is why most Clojure projects don't have such a setup sadly).

Failing tests can plausibly mean that _other_ things will be broken on windows so in a way, without a green build everywhere this PR will be of limited use.

Cheers - V

peterwang commented:
Hi, I have added detailed PR and inline comment for the code change.

Regarding to the FAIED tests on Windows, Please see the following snippets, thanks a lot.
```
>lein test
lein test
Compiling namespace compliment.context
Compiling namespace compliment.core
Compiling namespace compliment.sources
Compiling namespace compliment.sources.class-members
Compiling namespace compliment.sources.keywords
Compiling namespace compliment.sources.local-bindings
Compiling namespace compliment.sources.namespaces-and-classes
Compiling namespace compliment.sources.ns-mappings
Compiling namespace compliment.sources.resources
Compiling namespace compliment.sources.special-forms
Compiling namespace compliment.utils
WARNING: reader-conditional already refers to: #'clojure.core/reader-conditional in namespace: clojure.tools.reader.impl.utils, being replaced by: #'clojure.tools.reader.impl.utils/reader-conditional
WARNING: tagged-literal already refers to: #'clojure.core/tagged-literal in namespace: clojure.tools.reader.impl.utils, being replaced by: #'clojure.tools.reader.impl.utils/tagged-literal

lein test compliment.sources.t-class-members

lein test compliment.sources.t-keywords

lein test compliment.sources.t-local-bindings

lein test compliment.sources.t-namespaces-and-classes

lein test :only compliment.sources.t-namespaces-and-classes/ns-class-completion

FAIL in (ns-class-completion) (t_namespaces_and_classes.clj:30)
they are completed either according to the mapping in the given
  namespace, or by classpath scanning results
expected: (contains #{"clojure.stacktrace" "clojure.set" "clojure.java.shell" "clojure.string"} :gaps-ok)
  actual: ("clojure.spec.gen.alpha" "cloverage.source" "clojure.stacktrace" "clojure.core.server" "clojure.core.specs.alpha" "clojure.spec.alpha" "clojure.set" "clojure.string")

lein test :only compliment.sources.t-namespaces-and-classes/ns-class-completion

FAIL in (ns-class-completion) (t_namespaces_and_classes.clj:30)
they are completed either according to the mapping in the given
  namespace, or by classpath scanning results
expected: (contains #{{:candidate "clojure.java.shell", :type :namespace} {:candidate "clojure.java.browse", :type :namespace}} :gaps-ok)
  actual: ({:candidate "clojure.java.io", :type :namespace} {:candidate "clojure.java.io.Coercions", :type :class} {:candidate "clojure.java.io.IOFactory", :type :class} {:candidate "clojure.java.api.Clojure", :type :class})

lein test compliment.sources.t-ns-mappings

lein test :only compliment.sources.t-ns-mappings/vars-completion-test

FAIL in (vars-completion-test) (t_ns_mappings.clj:99)
extra metadata can be requested from this completion source
expected: (clojure.core/= [{:candidate "frequencies", :type :function, :ns "clojure.core", :arglists ["[coll]"], :doc "clojure.core/frequencies\n([coll])\n  Returns a map from distinct items in coll to the number of times\n  they appear.\n"}] (binding [*extra-metadata* #{:arglists :doc}] (doall (src/candidates "freq" (-ns) nil))))
  actual: (not (clojure.core/= [{:candidate "frequencies", :type :function, :ns "clojure.core", :arglists ["[coll]"], :doc "clojure.core/frequencies\n([coll])\n  Returns a map from distinct items in coll to the number of times\n  they appear.\n"}] ({:candidate "frequencies", :type :function, :ns "clojure.core", :arglists ("[coll]"), :doc "clojure.core/frequencies\r\n([coll])\r\n  Returns a map from distinct items in coll to the number of times\n  they appear.\r\n"})))

lein test compliment.sources.t-resources

lein test :only compliment.sources.t-resources/resources-test

FAIL in (resources-test) (t_resources.clj:10)
completion works when started from a string in a resource call
expected: (clojure.core/= [{:candidate "META-INF/maven/compliment/compliment/pom.properties", :type :resource}] (src/candidates "META" *ns* (ctx/parse-context (quote (resource "__prefix__")))))
  actual: (not (clojure.core/= [{:candidate "META-INF/maven/compliment/compliment/pom.properties", :type :resource}] ({:candidate "META-INF\\maven\\compliment\\compliment\\pom.properties", :type :resource})))

lein test :only compliment.sources.t-resources/resources-test

FAIL in (resources-test) (t_resources.clj:10)
completion works when started from a string in a resource call
expected: (clojure.core/= [{:candidate "META-INF/maven/compliment/compliment/pom.properties", :type :resource}] (src/candidates "META" *ns* (ctx/parse-context (quote (io/resource "__prefix__")))))
  actual: (not (clojure.core/= [{:candidate "META-INF/maven/compliment/compliment/pom.properties", :type :resource}] ({:candidate "META-INF\\maven\\compliment\\compliment\\pom.properties", :type :resource})))

lein test compliment.sources.t-special-forms

lein test compliment.t-benchmark
Evaluation count : 78 in 6 samples of 13 calls.
             Execution time mean : 8.517454 ms
    Execution time std-deviation : 176.648424 µs
   Execution time lower quantile : 8.309402 ms ( 2.5%)
   Execution time upper quantile : 8.690663 ms (97.5%)
                   Overhead used : 5.784280 ns

lein test compliment.t-context

lein test compliment.t-core

lein test :only compliment.t-core/completions-test

FAIL in (completions-test) (t_core.clj:85)
different metadata is attached to candidates
expected: (just [{:type :namespace, :candidate "clojure.test.tap"}])
  actual: ()

lein test compliment.t-utils

Ran 18 tests containing 196 assertions.
6 failures, 0 errors.
Tests failed.

```

vemv commented:
Thanks for the iteration! I'll see what I can make of the failing tests, will let you know.

As for the comment/desc, I'd expect to see what is being exactly fixed / what was exactly failing.

Saying "this fixes x" doesn't quite cut it because it doesn't show _why_.

vemv commented on src/compliment/utils.clj:
```diff
@@ -195,7 +195,10 @@
+                 (replace ".class" "") (replace File/separator ".")
+                 ;; Address the issue #79 , on Windows, for prefix such
+                 ;; as "java.util.", the list of candidates was empty.
+                 (replace "/" ".")))
```
Ok, so I now understand what you're doing: https://bitmingw.com/2020/05/16/do-not-use-file-separator-when-loading-java-resource-files/

In this case, looks like `(replace File/separator ".")` is redundant and should be removed.

vemv commented:
Inspecting https://github.com/alexander-yakushev/compliment/pull/79#issuecomment-916010711 :

Looks like `vars-completion-test` can be trivially upgraded from expecting `\n` to expecting something built off `(System/lineSeparator)`.

`resources-test` is legitimately failing. The `actual` value is `"META-INF\\maven\\compliment\\compliment\\pom.properties"`. That doesn't denote a valid resource name; those backslashes should be slashes, even on Windows.

So I'd suggest fixing the other `File/separator` usage that I can see in the codebase.

@peterwang how does that sound? Would be nice to fix at least a few tests.

peterwang commented:
Hi, it seems that FAILED tests on Windows were all caused by `File/separator`.  I've fixed the code and the coresponding tests. Regarding the `\n` problem, I've fixed the test following your suggestion of building the expected string from `System/lineSeparator`. Finally, all tests run on both Windows and Linux.

vemv commented on src/compliment/utils.clj:
```diff
@@ -219,5 +222,7 @@
+        ;; resource pathes always use "/" regardless of platform
+        (.. (if (.startsWith file File/separator)
+              (.substring file 1) file)
+            (replace File/separator "/"))))))
```
I'd suggest extracting `"/"` as:

```clj
(def resource-separator
  "The path separator used within resources and .jar files.

Note that should always have the same value, regardless of OS."
  "/")
```

vemv reviewed (approved):
🍻 🚀 great work!

Thanks for the time fixing the tests on Windows.

peterwang commented:
Done with `resource-separator`, thanks for your guidance.

vemv commented:
WDYT of these changes? @alexander-yakushev 

We're in the process of cutting a new cider-nrepl release, would love to include the current 2 PRs.
